File: Examples/Examples/Demo/WebViewController.m
//
//  WebViewController.m
//  Examples
//
//  Created by Victor Zhu on 2021/5/13.
//  Copyright © 2020 Airwallex. All rights reserved.
//

#import "WebViewController.h"
#import <WebKit/WebKit.h>

@interface WebViewController ()<WKNavigationDelegate, WKUIDelegate>
@property (nonatomic, strong) WKWebView *webView;

@end

@implementation WebViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    CGFloat SCREEN_WIDTH = [[UIScreen mainScreen] bounds].size.width;
    CGFloat SCREEN_HEIGHT = [[UIScreen mainScreen] bounds].size.height;
    WKWebViewConfiguration *config = [[WKWebViewConfiguration alloc] init];
    self.webView = [[WKWebView alloc] initWithFrame:CGRectMake(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT) configuration:config];
    [self.view addSubview:self.webView];
    [self.webView addObserver:self forKeyPath:@"title" options:(NSKeyValueObservingOptionNew)context:nil];
    [self load];
}

- (void)load {
    if (self.url.length) {
        NSURL *url = [NSURL URLWithString:self.url];
        NSMutableURLRequest *requeset = [NSMutableURLRequest requestWithURL:url cachePolicy:NSURLRequestReloadIgnoringCacheData timeoutInterval:10];
        NSURL *referUrl = [NSURL URLWithString:self.referer];
        if (referUrl) {
            NSString *host = referUrl.host;
            self.referer = [NSString stringWithFormat:@"%@://", host];
        }
        [requeset setValue:self.referer forHTTPHeaderField:@"Referer"];
        [requeset setValue:@"Airwallex-iOS-SDK" forHTTPHeaderField:@"User-Agent"];
        [requeset setHTTPMethod:@"GET"];
        self.webView.navigationDelegate = self;
        self.webView.UIDelegate = self;
        [self.webView loadRequest:requeset];
        //设置ua会造成微信支付 cancel 无法返回
        //        [self setWebViewUA];
    }
}

- (void)setWebViewUA {
    [self.webView evaluateJavaScript:@"navigator.userAgent"
                   completionHandler:^(id result, NSError *error) {
                       NSDictionary *dictionary = [NSDictionary dictionaryWithObjectsAndKeys:@"Airwallex-iOS-SDK", @"UserAgent", nil];
                       [[NSUserDefaults standardUserDefaults] registerDefaults:dictionary];
                       [[NSUserDefaults standardUserDefaults] synchronize];
                   }];
}

- (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigationAction *)navigationAction decisionHandler:(void (^)(WKNavigationActionPolicy))decisionHandler {
    NSString *absoluteString = navigationAction.request.URL.absoluteString;
    if (([absoluteString hasPrefix:@"weixin://wap/pay?"] ||
         [absoluteString hasPrefix:@"alipay://"] ||
         [absoluteString hasPrefix:@"alipayhk://"] ||
         [absoluteString hasPrefix:@"airwallexcheckout://"] ||
         [absoluteString hasPrefix:@"alipays://"]) &&
        !([absoluteString hasPrefix:@"http://"] || [absoluteString hasPrefix:@"https://"])) {
        NSURL *url = [[NSURL alloc] initWithString:absoluteString];
        if (url) {
            [[UIApplication sharedApplication] openURL:url options:@{} completionHandler:nil];
        }
        decisionHandler(WKNavigationActionPolicyAllow);
    } else {
        decisionHandler(WKNavigationActionPolicyAllow);
    }
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    if ([keyPath isEqualToString:@"title"]) {
        self.title = self.webView.title;
    }
}

@end

File: Examples/Examples/Controllers/WebViewController.swift
//
//  WebViewController.swift
//  Examples
//
//  Created by Weiping Li on 2025/2/12.
//  Copyright © 2025 Airwallex. All rights reserved.
//

import Combine
import UIKit
@preconcurrency import WebKit

class WebViewController: UIViewController {
    private var webView: WKWebView!

    let url: String
    private(set) var referer: String
    var isPopupWebView = false

    init(url: String, referer: String) {
        self.url = url
        self.referer = referer
        super.init(nibName: nil, bundle: nil)
    }

    /// Initialize with an external WKWebView (used for popup support, e.g., Google Pay)
    /// https://developers.google.com/pay/api/web/guides/recipes/using-ios-wkwebview
    init(webView: WKWebView) {
        self.url = ""
        self.referer = ""
        super.init(nibName: nil, bundle: nil)
        self.webView = webView
        self.webView.uiDelegate = self
        self.webView.navigationDelegate = self
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private var cancellable: AnyCancellable?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Create webView if not already provided (normal case vs popup case)
        if webView == nil {
            let config = WKWebViewConfiguration()
            // Append GOOGLE_PAY_SUPPORTED to user agent for Google Pay isReadyToPay API
            config.applicationNameForUserAgent = "Airwallex-iOS-SDK" + " GOOGLE_PAY_SUPPORTED"
            webView = WKWebView(frame: .zero, configuration: config)
            webView.uiDelegate = self
            webView.navigationDelegate = self
            if #available(iOS 16.4, *) {
                webView.isInspectable = true
            }
        }
        webView.translatesAutoresizingMaskIntoConstraints = false

        view.addSubview(webView)
        let constraints = [
            webView.topAnchor.constraint(equalTo: view.topAnchor),
            webView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            webView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            webView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
        ]
        NSLayoutConstraint.activate(constraints)

        cancellable = webView.publisher(for: \.title)
            .assign(to: \.title, on: self)

        if !url.isEmpty {
            load()
        }
    }

    private var webViewDidCloseHandled: Bool = false

    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        guard isPopupWebView && !webViewDidCloseHandled else { return }
        // Only execute when actually being dismissed/popped, not when covered by another VC
        guard isBeingDismissed || isMovingFromParent else { return }
        webViewDidCloseHandled = true
        // when dismissed or popped by user
        // force status sync (to dismiss GPay mask in original webView)
        webView.evaluateJavaScript("window.close()") { object, error in
            if let error {
                print(error)
            }
        }
    }

    func load() {
        guard !url.isEmpty, let requestUrl = URL(string: url) else { return }
        
        var request = URLRequest(url: requestUrl, cachePolicy: .reloadIgnoringCacheData, timeoutInterval: 10)
        
        if let referUrl = URL(string: referer), let host = referUrl.host {
            referer = "\(host)://"
        }
        
        request.setValue(referer, forHTTPHeaderField: "Referer")
        request.httpMethod = "GET"
        
        webView.load(request)
    }
}

extension WebViewController: WKNavigationDelegate {

    func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
        guard let absoluteString = navigationAction.request.url?.absoluteString else {
            decisionHandler(.allow)
            return
        }

        let customSchemes = ["weixin://wap/pay?", "alipay://", "alipayhk://", "airwallexcheckout://", "alipays://", "kakaotalk://"]

        if customSchemes.contains(where: { absoluteString.hasPrefix($0) }),
            let url = URL(string: absoluteString) {
            UIApplication.shared.open(url, options: [:], completionHandler: nil)
        }

        decisionHandler(.allow)
    }
}

// MARK: - WKUIDelegate (Popup support for Google Pay)
extension WebViewController: WKUIDelegate {

    func webView(
        _ webView: WKWebView,
        createWebViewWith configuration: WKWebViewConfiguration,
        for navigationAction: WKNavigationAction,
        windowFeatures: WKWindowFeatures
    ) -> WKWebView? {
        // Only handle popup requests (when targetFrame is nil)
        guard navigationAction.targetFrame == nil else { return nil }

        let popupWebView = WKWebView(frame: .zero, configuration: configuration)
        popupWebView.customUserAgent = "Airwallex-iOS-SDK" + " GOOGLE_PAY_SUPPORTED"

        let popupViewController = WebViewController(webView: popupWebView)
        popupViewController.isPopupWebView = true
        present(popupViewController, animated: true)

        return popupWebView
    }

    func webViewDidClose(_ webView: WKWebView) {
        guard !webViewDidCloseHandled else { return }
        webViewDidCloseHandled = true
        if let presentingViewController {
            presentingViewController.dismiss(animated: true)
        } else if let navigationController {
            navigationController.popViewController(animated: true)
        }
    }
}
